![alt text][logo]

[logo]:https://github.com/NCBI-Hackathons/DiseaseCluster/blob/master/DC-logo-wide.png

### DiseaseCluster is a pipeline of disease prediction based on transcriptomic clustering
#### Hackathon Team
Lead - Mathew Moss

SysAdmins - Gordon Lemmon, Yaxin Xue

Data Wrangler - Ben Harris

PseudoBulk Pipline, Data Finder, Writer - Meghan Ferrall-Fairbanks

### How to cite this work in a publication: [citation to be determined here]
[doi here]

In the move towards personalized medicine, it would be helpful if physicians could visualize their patients across different disease states to help determine potential therapeutic options for their patients. As wet-lab scientific techniques move from the bench-to-bedside, RNA-sequencing can allow physicians the opportunity to detect and quantify patient-specific mRNA signature and understand more about how they relate to cellular signaling in a disease-context. 

*DiseaseCluster* is a versatile workflow that using uniform manifold approximation and projection (UMAP) to clusters transcriptomic data for a variety of disease states and allows researchers and clinicians to visually compare an uploaded patient sample in the disease space. 

Objective: Create a reusable, reproducible, and interactive workflow to cluster transcriptomic data based on disease state. Furthermore, the workflow will allow for researchers/clinicians to upload new transcriptomic datasets to visualize how the new data compares to the disease space. The disease space currently describes cancer samples profiled in TCGA. 

This project was part of the [2018 Cold Spring Harbor Biological Data Science NCBI Hackathon](https://biohackathons.github.io/).

## Dependencies

The pipeline was developed in [Python3](https://www.python.org/download/releases/3.0/) with the following library/package dependencies:
* [scikit-learn](https://scikit-learn.org/stable/)
* [pandas](https://pandas.pydata.org/)
* [numpy](http://www.numpy.org/)
* [scipy](https://www.scipy.org/)
* [Bokeh](https://bokeh.pydata.org/en/latest/)
* [Seaborn](https://seaborn.pydata.org/)
* [Matplotlib](https://matplotlib.org/)
* [UMAP](https://github.com/lmcinnes/umap)

Exploratory analysis of converting single cell RNA-seq data to pseudo-bulk RNA-seq for comparitive analysis was written in [Julia v0.7](https://julialang.org/).

## DiseaseCluster Workflow

![alt text][pipeline]

[pipeline]:https://github.com/NCBI-Hackathons/DiseaseCluster/blob/master/PipelineFig2.png "DiseaseCluster Workflow"

The DiseaseCluster pipeline workflow is described in the figure above. First, TCGA transcriptomic data was downloaded and additional bulk RNA-seq data will be added to expand disease types analyzed. Data was converted from FPKM to TPM, non-protein coding genes were removed and the data was log transformed. UMAP (uniform manifold approximation and projection) was performed on the expression matrix to reduce the dimensionality to two dimensions in latent space based on gene expression. Clustered data was then overlaid with annotated information about the disease state, including known variants. This map is then visualized by a boken interactive user interface.

## Installation

More information coming soon about installation 

## Input File Format and Usage

The input to the pipeline is RNA-seq data across a wide range of diseases. Currently the input data only contains cancer disease RNA-seq values from the TCGA. This can be extended to include other disease by concatenating columns from other bulk RNA-seq datasets (see Future Directions section below).  

Researchers and clinicians can "add one more" and input new RNA-seq sample to see how that sample would map onto the disease space. To add one more, the pipeline accepts a two column text file with gene names in the first column and RNA-seq counts in the second column (no header). Gene names will be coverted to Ensembl ID, if given in gene symbol format (note: Ensembl ID is preferred for input).  

## Output

The output of the initial pipeline is a boken interface that allows the user to interact with the clustered RNA-seq disease space data.


The results shown here are in whole or part based upon data generated by the TCGA Research Network: http://cancergenome.nih.gov/.

## Exploration and Validation

#### Pseudo-bulk RNA-seq Test
Goal: Project new patient/sample information on the disease space to see where that sample would cluster. However, many new experiments utilize single cell RNA-sequencing (scRNAseq), so we need to convert this data to pseudo-bulk RNA-seq for comparison in our pipeline.

Here we take scRNAseq from 6 different triple negative breast cancer (TNBC) patients, convert to pseudo-bulk data by comparing a few different mean calculations of the single cell RNA-seq counts per person, then see how/where these individual patients cluster in the overall disease space. 

The code to create the pseudo-bulk RNA-seq was written in Julia and is available in the Jupyter Notebook [here](https://github.com/NCBI-Hackathons/DiseaseCluster/blob/master/Pseduo-bulk-from-scRNAseq.ipynb). The dataset was the TPM dataset available from GEO accession number [GSE118389](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118389). This dataset contains single cell RNA sequencing of 1,534 cells from six fresh triple negative breast cancer tumors. The jupyter notebook divides the overall dataset up by patient, then calculates the (1) arithmetic mean, (2) geometric mean, (3) harmonic mean, (4) power mean, and (5) naive summation over counts, to achieve a pseudo-bulk characterization to describe the single cells for an individual patient. For the geometric and harmonic means, only cells with positive TPM values were used in the mean calculation. Finally, the pseudo-bulk RNA-seq can be saved using the savedata() function in the notebook, which could then be imported into the pipeline. 

## Future Directions: Expanding Disease Space 

To further extend this tool, it would be helpful to increase the number of diseases represented in the initial disease space RNA-seq pipeline file. TCGA data has on the order of 100's of samples per cancer type, so larger cohorts of different disease phenotypes should be added at a time to obtain reasonable clustering. (i.e. if you just add in one diabetes patient, that sample would likely not cluster alone by itself)

Other datasets include:
* Healthy/Normal Individuals, GEO: [GSE36192](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36192), from 396 patients
* Alzheimer's Disease Individuals, GEO: [GSE95587](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE95587), from 117 patients
* HIV-infected Individuals, GEO: [GSE115449](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115449), from 95 individuals
* Liver Disease Patients, EBI: [E-MTAB-6863](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-6863/), from 69 patients

## F.A.Q. 
1. How to cite? 

Citation information coming soon

2. How to use? 

Follow the instructions on this page.

3. What if I need help?

Feel free to contact authors if you need help. 

## References

Karaayvaz M, Cristea S, Gillespie SM, Patel AP et al. Unravelling subclonal heterogeneity and aggressive disease states in TNBC through single-cell RNA-seq. Nat Commun 2018 Sep 4;9(1):3588. PMID: [30181541](https://www.ncbi.nlm.nih.gov/pubmed/30181541)

McInnes, L, Healy, J, UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction, [ArXiv e-prints 1802.03426](https://arxiv.org/abs/1802.03426), 2018

Weinstein JN, Collisson EA, Mills GB, Shaw KM et al. The Cancer Genome Atlas Pan-Cancer Analysis Project. Nat Genet 2013 Oct; 45(10):1113-1120. PMID: [24071849](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3919969/)

## People/Team
* **Matthew Moss**, CSHL, Cold Spring Harbor, NY, USA, [moss@cshl.edu](mailto:moss@cshl.edu) 
* **Meghan Ferrall-Fairbanks**, Moffitt Cancer Center and Research Institute, Tampa, FL, USA, [meghan.ferrall-fairbanks@moffitt.org](mailto:meghan.ferrall-fairbanks@moffitt.org) 
* **Benjamin Harris**, CSHL, Cold Spring Harbor, NY, USA, [bharris@cshl.edu](mailto:bharris@cshl.edu)
* **Gordon Howard Lemmon**, University of Utah, Salt Lake City, UT, USA, [gordon.lemmon@utah.edu](mailto:gordon.lemmon@utah.edu)
* **Yaxin Xue**, University of Bergen, Bergen, Norwary, [yaxin.xue@uib.no](mailto:yaxin.xue@uib.no)


